File: packages/cli-clients/node/contextdesk_client.mjs
#!/usr/bin/env node
/**
 * Minimal Node/TypeScript-friendly thin client for contextdesk.
 * Uses child_process.spawn with argv array — never shell:true.
 * See docs/CLI_CLIENT_PROTOCOL.md.
 */
import { spawn } from "node:child_process";
import process from "node:process";

function resolveBin() {
  return process.env.CONTEXTDESK_BIN || "contextdesk";
}

const completedVerdicts = new Map([[8, "not_ready"], [9, "non_conforming"], [10, "partial"]]);
const MAX_ENVELOPE_BYTES = 1024 * 1024;

/**
 * @param {string[]} args command args after global flags
 * @param {{ dataDir?: string, env?: NodeJS.ProcessEnv, cwd?: string }} [opts]
 */
export function runJson(args, opts = {}) {
  const bin = resolveBin();
  const argv = [];
  if (opts.dataDir) argv.push("--data-dir", opts.dataDir);
  argv.push("--json", ...args);

  return new Promise((resolve, reject) => {
    const child = spawn(bin, argv, {
      env: { ...process.env, ...(opts.env || {}) },
      cwd: opts.cwd,
      shell: false, // critical: no shell command construction
      stdio: ["ignore", "pipe", "pipe"],
    });
    let stdout = "";
    let stderr = "";
    child.stdout.on("data", (c) => (stdout += c));
    child.stderr.on("data", (c) => (stderr += c));
    child.on("error", reject);
    child.on("close", (code) => {
      const line = stdout.trim();
      if (Buffer.byteLength(line, "utf8") > MAX_ENVELOPE_BYTES) {
        reject(new Error(`oversized JSON envelope (exit=${code})`));
        return;
      }
      let envelope;
      try {
        envelope = JSON.parse(line);
      } catch (e) {
        reject(new Error(`invalid JSON (exit=${code}): ${stderr || line}`));
        return;
      }
      if (!envelope || Array.isArray(envelope) || typeof envelope !== "object" || typeof envelope.ok !== "boolean") {
        reject(new Error(`envelope requires top-level boolean ok (exit=${code})`));
        return;
      }
      if (envelope.ok === false) {
        const err = envelope.error || {};
        const e = new Error(err.message || "command failed");
        e.kind = err.kind || "internal";
        e.exitCode = code;
        reject(e);
        return;
      }
      if (code !== 0 && !completedVerdicts.has(code)) {
        reject(Object.assign(new Error(`ok envelope but exit ${code}`), { kind: "internal", exitCode: code }));
        return;
      }
      resolve({ envelope, exitCode: code, verdictKind: completedVerdicts.get(code) || null });
    });
  });
}

export function capabilities(opts) {
  return runJson(["capabilities"], opts);
}

const isMain = process.argv[1] && process.argv[1].endsWith("contextdesk_client.mjs");
if (isMain) {
  const args = process.argv.slice(2);
  let dataDir;
  const cmd = [];
  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--data-dir") {
      dataDir = args[++i];
    } else {
      cmd.push(args[i]);
    }
  }
  if (cmd.length === 0) cmd.push("capabilities");
  runJson(cmd, { dataDir })
    .then((result) => {
      console.log(JSON.stringify(result.envelope, null, 2));
      process.exitCode = result.exitCode;
    })
    .catch((e) => {
      console.error(JSON.stringify({ ok: false, error: { kind: e.kind || "internal", message: e.message } }));
      process.exit(e.exitCode || 70);
    });
}

File: packages/cli-clients/c/contextdesk_client.c
/* Minimal C thin client — execvp argv array only (no system()).
 * Direct C ABI into cd-core is a staged FUTURE option, not this path.
 * See docs/CLI_CLIENT_PROTOCOL.md.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>

static const char *resolve_bin(void) {
    const char *b = getenv("CONTEXTDESK_BIN");
    return (b && b[0]) ? b : "contextdesk";
}

typedef enum {
    CONTEXTDESK_NO_VERDICT = 0,
    CONTEXTDESK_NOT_READY,
    CONTEXTDESK_NON_CONFORMING,
    CONTEXTDESK_PARTIAL
} contextdesk_verdict;

/* Use only after parsing an ok:true envelope. Stable kinds are not_ready,
 * non_conforming, and partial: report-bearing verdicts, never internal. */
contextdesk_verdict contextdesk_completed_verdict(int exit_code) {
    if (exit_code == 8) return CONTEXTDESK_NOT_READY;
    if (exit_code == 9) return CONTEXTDESK_NON_CONFORMING;
    if (exit_code == 10) return CONTEXTDESK_PARTIAL;
    return CONTEXTDESK_NO_VERDICT;
}

/* argv: ["capabilities"] etc. Prints child stdout; returns exit code. */
int contextdesk_run_json(const char *data_dir, char *const cmd_args[], int cmd_argc) {
    const char *bin = resolve_bin();
    /* max 64 argv slots */
    char *argv[64];
    int n = 0;
    argv[n++] = (char *)bin;
    if (data_dir && data_dir[0]) {
        argv[n++] = "--data-dir";
        argv[n++] = (char *)data_dir;
    }
    argv[n++] = "--json";
    for (int i = 0; i < cmd_argc && n < 62; i++) {
        argv[n++] = cmd_args[i];
    }
    argv[n] = NULL;

    pid_t pid = fork();
    if (pid < 0) {
        perror("fork");
        return 70;
    }
    if (pid == 0) {
        execvp(bin, argv);
        perror("execvp");
        _exit(127);
    }
    int status = 0;
    if (waitpid(pid, &status, 0) < 0) {
        perror("waitpid");
        return 70;
    }
    if (WIFEXITED(status)) {
        return WEXITSTATUS(status);
    }
    return 130; /* signalled ~ cancelled */
}

int main(int argc, char **argv) {
    const char *data_dir = NULL;
    char *cmd[32];
    int cmd_n = 0;
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--data-dir") == 0 && i + 1 < argc) {
            data_dir = argv[++i];
        } else if (cmd_n < 31) {
            cmd[cmd_n++] = argv[i];
        }
    }
    if (cmd_n == 0) {
        cmd[cmd_n++] = "capabilities";
    }
    cmd[cmd_n] = NULL;
    /* Parent does not capture pipes in this minimal example — child inherits stdout.
     * Production C clients should pipe stdout and parse JSON envelopes. */
    return contextdesk_run_json(data_dir, cmd, cmd_n);
}
